### PR TITLE
(WIP) test-interface

### PR DESCRIPTION
**Pull request by MasseGuillaume**
* creates a separate sbt project to run tests: foo => foo-native-test-main
* fix infinite loop in String.indexOf(String, Int): Int
* utest interface
* succesfuly ran utest:
      see: https://github.com/scala-native-ports/utest/pull/1
      sbt sandbox-native-test-main/test

@Duhemm

### CODE REVIEW

File: bin/nix-run [hunk 1]
```diff
@@ -1,3 +1,5 @@
 #!/usr/bin/env bash
 
-nix-shell bin/scala-native.nix -A clangEnv
+HERE="`dirname $0`"
+
+nix-shell $HERE/scala-native.nix -A clangEnv
```
Comment (densh) [inline bin/nix-run]: This should go into separate PR.

File: build.sbt [hunk 2]
```diff
@@ -360,3 +377,31 @@ lazy val benchmarks =
...
+  Command.command("testAll") { state =>
+    "tools/test" ::
+      "sandbox/run" ::
+        "demoNative/run" ::
+          "tests/run" ::
+            "benchmarks/run" ::
+              "scripted" ::
+                state
+  }
+)
```
Comment (densh) [inline build.sbt]: @sjrd @Duhemm is it the nicest way to aggregate tasks? Can't we just redefine `test`, `publishLocal` and `clean` on the root project?

File: nativelib/src/main/scala/java/lang/Class.scala [hunk 3]
```diff
@@ -6,7 +6,7 @@ import scala.scalanative.runtime.Type
 final class _Class[A](val ty: Ptr[Type]) {
   def getName(): String = (!ty).name
 
-  override def hashCode: Int = ty.cast[Long].##
+  override def hashCode: Int = getName().##
```
Comment (densh) [inline nativelib/src/main/scala/java/lang/Class.scala]: Caused by https://github.com/scala-native/scala-native/issues/404

File: sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala [hunk 4]
```diff
@@ -169,18 +169,67 @@ object ScalaNativePluginInternal {
...
     addCompilerPlugin(
       "org.scala-native" % "nscplugin" % nativeVersion cross CrossVersion.full),
     nativeVerbose := false,
     nativeEmitDependencyGraphPath := None,
     nativeLibraryLinkage := Map(),
-    nativeSharedLibrary := false,
+    nativeSharedLibrary := false
+  )
+
+  def derivedProjects(proj: ProjectDefinition[_]): Seq[Project] =
```
Comment (densh) [inline sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala]: Needs a detailed comment to explain what's going on down there.

File: sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/TestUtilities.scala [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        |    ${cases.mkString(System.lineSeparator + indent)}
+        |    case Array() =>
+        |      ()
+        |  }
+        |}""".stripMargin
+  }
+
+  private def createTestCase(test: TestDefinition): String = {
+    val body =
+      test.fingerprint match {
```
Comment (densh) [inline sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/TestUtilities.scala]: Will this work for other frameworks in the future like ScalaTest/ScalaCheck/etc?

**Comment by densh** [comment]
Currently crashes in `java.util.concurrent.TimeUnit::convert_i64_class.java.util.concurrent.TimeUnit_i64` with: 

```
│* thread #1: tid = 0x4b8862, 0x000000010002bffa tests-out`java.util.concurrent.TimeUnit::convert_i64_class.java.util.concurrent.TimeUnit_i64 + 42, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x8)
│  * frame #0: 0x000000010002bffa tests-out`java.util.concurrent.TimeUnit::convert_i64_class.java.util.concurrent.TimeUnit_i64 + 42
│    frame #1: 0x000000010002c170 tests-out`java.util.concurrent.TimeUnitSuite$$anonfun$1::apply$mcV$sp_unit + 48
│    frame #2: 0x00000001000baf75 tests-out`tests.Suite$$anonfun$test$1::apply$mcZ$sp_bool + 53
│    frame #3: 0x00000001000bae85 tests-out`tests.Suite$$anonfun$run$1::apply_class.tests.Test_unit + 53
│    frame #4: 0x00000001000bae36 tests-out`tests.Suite$$anonfun$run$1::apply_class.java.lang.Object_class.java.lang.Object + 6
│    frame #5: 0x00000001000a7ce9 tests-out`scala.collection.mutable.UnrolledBuffer$Unrolled::foreach_trait.scala.Function1_unit + 217
│    frame #6: 0x00000001000a83ba tests-out`scala.collection.mutable.UnrolledBuffer::foreach_trait.scala.Function1_unit + 10
│    frame #7: 0x00000001000bb49b tests-out`tests.Suite::run_bool + 155
│    frame #8: 0x00000001000bacf1 tests-out`tests.Main$$anonfun$main$1::apply_class.tests.Suite_unit + 17
│    frame #9: 0x00000001000bacc6 tests-out`tests.Main$$anonfun$main$1::apply_class.java.lang.Object_class.java.lang.Object + 6
│    frame #10: 0x00000001000688b2 tests-out`scala.collection.immutable.List::foreach_trait.scala.Function1_unit + 178
│    frame #11: 0x00000001000bade7 tests-out`tests.Main$::main_class.ssnr.ObjectArray_unit + 135
│    frame #12: 0x00000001000c19ed tests-out`main + 61
│    frame #13: 0x00007fffe4614255 libdyld.dylib`start + 1
```

**Comment by MasseGuillaume** [comment]
@densh how you got that stacktrace? lldb ?

**Comment by densh** [comment]
Yep, you can do `lldb mybinary` and then `run` and `bt`.

**Comment by densh** [comment]
This one also needs a rebase.

File: nativelib/src/main/scala/java/lang/Class.scala (hunk 3 above)
Comment (densh) [inline nativelib/src/main/scala/java/lang/Class.scala]: Fixed in #421

Comment (densh) [inline nativelib/src/main/scala/java/lang/Class.scala]: The fix has been merged, this change is not needed any longer.

**Comment by densh** [comment]
We need to split this one into smaller stand-alone PRs.

File: javalib/src/main/scala/java/util/concurrent/TimeUnit.scala [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+  protected def makeNext: Long => Long
+  protected def makePrev: Long => Long
+  protected def next: TimeUnit
+  protected def prev: TimeUnit
+
+  def convert(sourceDuration: Long, sourceUnit: TimeUnit): Long =
+    if (id > sourceUnit.id) convert(makeNext(sourceDuration), sourceUnit.next)
+    else if (id == sourceUnit.id) sourceDuration
+    else convert(makePrev(sourceDuration), sourceUnit.prev)
```
Comment (densh) [inline javalib/src/main/scala/java/util/concurrent/TimeUnit.scala]: There is an NPE somewhere here. As discovered by @MasseGuillaume.

Comment (densh) [inline javalib/src/main/scala/java/util/concurrent/TimeUnit.scala]: Another data point is an implementation from Scala.js: https://github.com/scala-js/scala-js/blob/master/javalib/src/main/scala/java/util/concurrent/TimeUnit.scala
